packages/babel-parser/src/plugins/typescript.js

```diff
@@ -319,6 +319,30 @@ export default (superClass: Class<Parser>): Class<Parser> =>
       return this.finishNode(node, "TSTypeParameterDeclaration");
     }
 
+    tsTryNextParseConstantContext(): ?N.TsTypeReference {
+      if (this.lookahead().type === tt._const) {
+        this.next();
+        return this.tsParseTypeReference();
+      }
+      return null;
+    }
+
+    tsCheckLiteralForConstantContext(node: N.Node) {
+      switch (node.type) {
+        case "StringLiteral":
+        case "TemplateLiteral":
+        case "NumericLiteral":
+        case "BooleanLiteral":
+        case "ObjectExpression":
+        case "ArrayExpression":
```

Comment:
For arrays and objects, don't we need to check their children?

e.g.
```js
[ foo() ] as const
```

Reply:
updated and added new test case